File: web/src/lib/types/task/lifecycle.ts
import { z } from 'zod/v4';

export const LifecycleTaskSchema = z.object({
    id: z.number().int(),
    guestType: z.enum(['vm', 'jail', 'jail-template', 'vm-template']),
    guestId: z.number().int(),
    action: z.string(),
    source: z.string(),
    status: z.enum(['queued', 'running', 'success', 'failed']),
    requestedBy: z.string().nullable().optional(),
    message: z.string().nullable().optional(),
    error: z.string().nullable().optional(),
    overrideRequested: z.boolean().default(false),
    startedAt: z.string().nullable().optional(),
    finishedAt: z.string().nullable().optional(),
    createdAt: z.string(),
    updatedAt: z.string()
});

export type LifecycleTask = z.infer<typeof LifecycleTaskSchema>;

const activeLifecycleTaskStatuses = new Set<LifecycleTask['status']>(['queued', 'running']);

export function isLifecycleTaskActive(task: LifecycleTask | null | undefined): task is LifecycleTask {
    if (!task) {
        return false;
    }

    if (!activeLifecycleTaskStatuses.has(task.status)) {
        return false;
    }

    return task.action.trim().length > 0;
}

File: internal/db/models/task/lifecycle.go
// SPDX-License-Identifier: BSD-2-Clause
//
// Copyright (c) 2025 The FreeBSD Foundation.
//
// This software was developed by Hayzam Sherif <hayzam@alchemilla.io>
// of Alchemilla Ventures Pvt. Ltd. <hello@alchemilla.io>,
// under sponsorship from the FreeBSD Foundation.

package taskModels

import "time"

const (
	GuestTypeVM           = "vm"
	GuestTypeJail         = "jail"
	GuestTypeJailTemplate = "jail-template"
	GuestTypeVMTemplate   = "vm-template"
)

const (
	LifecycleTaskStatusQueued  = "queued"
	LifecycleTaskStatusRunning = "running"
	LifecycleTaskStatusSuccess = "success"
	LifecycleTaskStatusFailed  = "failed"
)

const (
	LifecycleTaskSourceUser    = "user"
	LifecycleTaskSourceStartup = "startup"
)

type GuestLifecycleTask struct {
	ID uint `gorm:"primaryKey" json:"id"`

	GuestType string `gorm:"index;not null" json:"guestType"`
	GuestID   uint   `gorm:"index;not null" json:"guestId"`
	Action    string `gorm:"index;not null" json:"action"`
	Source    string `gorm:"index;not null;default:user" json:"source"`

	Status string `gorm:"index;not null;default:queued" json:"status"`

	RequestedBy string `json:"requestedBy"`
	Message     string `json:"message"`
	Error       string `gorm:"type:text" json:"error"`
	Payload     string `gorm:"type:text" json:"payload"`

	OverrideRequested bool `gorm:"index;default:false" json:"overrideRequested"`

	StartedAt  *time.Time `json:"startedAt"`
	FinishedAt *time.Time `json:"finishedAt"`

	CreatedAt time.Time `gorm:"autoCreateTime" json:"createdAt"`
	UpdatedAt time.Time `gorm:"autoUpdateTime" json:"updatedAt"`
}

func (GuestLifecycleTask) TableName() string {
	return "guest_lifecycle_tasks"
}
